amivapi/auth/link_methods.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+#
+# license: AGPLv3, see LICENSE for details. In addition we strongly encourage
+#          you to buy us beer if we meet and you like the software.
+
+"""Functions adding permitted methods to '_links' part of response.
+
+Links are only added for resources using AmivTokenAuth.
+"""
+
+import json
+
+from flask import current_app, g
+from eve.auth import resource_auth
+
+from .auth import AmivTokenAuth, authenticate, check_if_admin
+
+
+def _get_item_methods(resource, item):
+    res = current_app.config['DOMAIN'][resource]
+    user = g.get('current_user')  # TODO: post_internal_problem
+    auth = resource_auth(resource)
+    is_admin = g.get('resource_admin')
+
+    # If the item is displayed, the read methods are obviously allowed
+    methods = ['GET', 'HEAD', 'OPTIONS'] + res['public_item_methods']
+
+    # Admins have access to all methods. For non admins check user permission.
+    if is_admin or auth.has_item_write_permission(user, item):
+        methods += res['item_methods']
+
+    # Remove duplicates before returning
+    return list(set(methods))
+
+
+def _get_resource_methods(resource):
+    res = current_app.config['DOMAIN'][resource]
+    auth = resource_auth(resource)
+    user = g.get('current_user')
+    is_admin = g.get('resource_admin')
+
+    # public methods and OPTIONS are always available
+    methods = res['public_methods'] + ['OPTIONS']
+    if 'GET' in methods:
+        methods += ['HEAD']
+
+    # resources may not have public read access, but we still can see the
+    # resource on the home endpoint
+    if user or is_admin or g.get('resource_admin_readonly'):
+        methods += ['GET', 'HEAD']
+
+    # write methods
+    if is_admin or auth.has_resource_write_permission(user):
+        methods += res['resource_methods']
+
+    # Remove duplicates
+    return list(set(methods))
+
+
+def _home_link_methods():
+    return ['GET', 'HEAD', 'OPTIONS']
+
+
+def add_methods_to_item_links(resource, item):
+    """Add methods to all links of the item.
+
+    Args:
+        resource (str): The name of the resource
+        item (dict): The item, must have a 'links' key.
+    """
+    links = item['_links']
+
+    # self
+    links['self']['methods'] = _get_item_methods(resource, item)
+
+    # parent, i.e. home -> only read methods (optional)
+    if 'parent' in links:
+        links['parent']['methods'] = _home_link_methods()
+
+    # collection -> resource (optional)
+    if 'collection' in links:
+        links['collection']['methods'] = _get_resource_methods(resource)
+
+
+def add_methods_to_resource_links(resource, response):
+    """Add methods to all links of the item.
```

Comment:
This looks like a copy-paste ;)
It is unclear to me where the difference to the above method lays. 
If I understand corrrectly, this here only adds links to a response?